### PR TITLE
Use time.perf_counter instead of time.time

### PR DESCRIPTION
**Pull request by sumanthratna**
## Why are these changes needed?

`time.perf_counter` should be used for measuring how long a block of code takes, not `time.time`. 

## Related issue number

n/a, see #8686 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
yes, but there were version conflicts so it may not have formatted all files
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
n/a
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/state.py [hunk 1]
```diff
@@ -637,7 +637,7 @@ def _job_length(self):
             overall_largest = max(overall_largest, rev_range[0][1])
 
             num_tasks += self.redis_client.zcount(
-                event_log_set, min=0, max=time.time())
+                event_log_set, min=0, max=time.perf_counter())
```
Comment (sumanthratna) [inline python/ray/state.py:640]: I'm not sure about this.

File: rllib/tests/test_execution.py [hunk 2]
```diff
@@ -109,12 +109,12 @@ def test_metrics(ray_start_regular_shared):
...
         })
 
-    start = time.time()
+    start = time.perf_counter()
     res1 = next(b)
     assert res1["episode_reward_mean"] > 0, res1
     res2 = next(b)
     assert res2["episode_reward_mean"] > 0, res2
-    assert time.time() - start > 2.4
+    assert time.perf_counter() - start > 2.4
```
Comment (sumanthratna) [inline rllib/tests/test_execution.py:117]: Hopefully this doesn't break anything; let me know if `time.time` should be used here.

File: rllib/optimizers/async_samples_optimizer.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def __init__(self,
 
         # Stats
         self._optimizer_step_timer = TimerStat()
-        self._stats_start_time = time.time()
+        self._stats_start_time = time.perf_counter()
```
Comment (sumanthratna) [inline rllib/optimizers/async_samples_optimizer.py:82]: I'm not sure about this; is `self. _stats_start_time` used anywhere else, where the actual time is needed, as opposed to the time taken?

File: rllib/execution/metric_ops.py [hunk 4]
```diff
@@ -118,9 +118,9 @@ class OncePerTimeInterval:
 
     Examples:
         >>> throttled_op = train_op.filter(OncePerTimeInterval(5))
-        >>> start = time.time()
+        >>> start = time.perf_counter()
         >>> next(throttled_op)
-        >>> print(time.time() - start)
+        >>> print(time.perf_counter() - start)
```
Comment (sumanthratna) [inline rllib/execution/metric_ops.py:123]: I'm not sure if the comment on the output of this `print` statement still applies.

**Comment by richardliaw** [comment]
Hey @sumanthratna, thanks for opening this PR. What is the benefit of this change?

**Comment by rkooo567** [comment]
Do you happened to know how much overhead perf_counter has compared to time.time?

**Comment by sumanthratna** [comment]
> Do you happened to know how much overhead perf_counter has compared to time.time?

Ah, I didn't think of that. This might not be the best way to benchmark:
`python3 -m timeit -vvvvv 'time.time() - time.time()'`:
```
1 loop -> 2.727e-06 secs
2 loops -> 1.752e-06 secs
5 loops -> 2.027e-06 secs
10 loops -> 3.599e-06 secs
20 loops -> 6.557e-06 secs
50 loops -> 1.5974e-05 secs
100 loops -> 3.1714e-05 secs
200 loops -> 0.000112558 secs
500 loops -> 0.000273654 secs
1000 loops -> 0.000450342 secs
2000 loops -> 0.000876811 secs
5000 loops -> 0.002070866 secs
10000 loops -> 0.003144788 secs
20000 loops -> 0.006614945 secs
50000 loops -> 0.01485118 secs
100000 loops -> 0.03512047 secs
200000 loops -> 0.06350689 secs
500000 loops -> 0.1510121 secs
1000000 loops -> 0.3064715 secs

raw times: 299.5733 msec, 298.0796 msec, 299.7568 msec, 301.0025 msec, 297.3337 msec

1000000 loops, best of 5: 297.3337 nsec per loop
```

`python3 -m timeit -vvvvv 'time.perf_counter() - time.perf_counter()'`:
```
1 loop -> 1.852e-06 secs
2 loops -> 1.36e-06 secs
5 loops -> 2.162e-06 secs
10 loops -> 3.838e-06 secs
20 loops -> 7.513e-06 secs
50 loops -> 1.8328e-05 secs
100 loops -> 3.6377e-05 secs
200 loops -> 9.8468e-05 secs
500 loops -> 0.000221442 secs
1000 loops -> 0.000388951 secs
2000 loops -> 0.000722758 secs
5000 loops -> 0.00178409 secs
10000 loops -> 0.004355535 secs
20000 loops -> 0.00927448 secs
50000 loops -> 0.01722481 secs
100000 loops -> 0.03992531 secs
200000 loops -> 0.07034141 secs
500000 loops -> 0.1730116 secs
1000000 loops -> 0.343489 secs

raw times: 350.2345 msec, 345.4968 msec, 340.3085 msec, 348.0912 msec, 350.3467 msec

1000000 loops, best of 5: 340.3085 nsec per loop
```

**Pull request closed without merging**